### PR TITLE
Ft cpdt 112

### PR DESCRIPTION
**Pull request by mlear**
The shuttle component.

### CODE REVIEW

**Comment by ghost-** [comment]
https://rawgit.com/fanappics/buoy/ft-cpdt-112/docs/index.html#bshuttle

File: src/components/BSelect/BSelect.vue [hunk 1]
```diff
@@ -69,7 +69,7 @@ export default {
       required: true
     },
     /**
-    * Array of objects {id: number, displayName: string} to be displayed as options.
+    * Array of objects {id: number, displayText: string} to be displayed as options.
```
Comment (ghost-) [inline src/components/BSelect/BSelect.vue]: Were these renamed just to be consistent?

File: src/components/BShuttle/BShuttle.test.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
+import { shallow } from 'vue-test-utils'
```
Comment (ghost-) [inline src/components/BShuttle/BShuttle.test.js]: Does shallow give you enough detail about the child elements, or would mount be better.

**Comment by ghost-** [comment]
One question.  Is there any validation needed for this component.  Thinks like a certain number of options must be selected?

File: src/components/BSelect/BSelect.vue (hunk 1 above)
Comment (mlear) [reply src/components/BSelect/BSelect.vue]: yes

File: src/components/BShuttle/BShuttle.test.js (hunk 2 above)
Comment (mlear) [reply src/components/BShuttle/BShuttle.test.js]: there are now child components being rendered, so mount isn't needed

File: src/components/BShuttle/BShuttle.vue [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+        type='button'
+        >
+        All <i class='fa fa-arrow-right' aria-hidden='true'></i>
+      </button>
+      <button 
+        @click="onMoveSelectedClick(selectedOptions, 'available')"
+        id='selected-to-chosen'
+        type='button'
+        >
+        <i class='fa fa-arrow-right' aria-hidden='true'></i>
```
Comment (ghost-) [inline src/components/BShuttle/BShuttle.vue]: These aren't showing up on the style guide.  I assume font awesome needs to be added to the /src/styleguide.html?  If font awesome is going to be a dependency we should look into getting it included in the buoy build.

**Comment by mlear** [comment]
im not sure about validation, there was nothing in the user story about whether or not this component was ever going to be "required" or have some sort of minimum/maximum selection

File: src/components/BShuttle/BShuttle.vue (hunk 3 above)
Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: nope, im waiting for @thedevelopnik 's pr that includes ionicons, and then I will switch them out once I get the new styles from Steve

**Comment by ghost-** [comment]
Functionally it seems to work well.  I even tested out the keyboard bindings.  Good work.  Can't wait to see it styled.

File: src/components/BShuttle/BShuttle.vue [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+    <div class='available'>
+      <label id='available-label'>{{ availableLabel }}</label>
+      <ul
+        aria-labeledby='available-label'
+        role='selectbox'
+      >
+        <li
+          v-if="placeholder && (!options || options.length === 0)"
+          :aria-label="placeholder"
+          class='placeholder'
```
Comment (eleanormal) [inline src/components/BShuttle/BShuttle.vue]: This or the `ul` needs `tabindex="0"`. I'll test it after standup to figure out which one.

File: src/components/BShuttle/BShuttle.vue [hunk 5]
```diff
@@ -0,0 +1,404 @@
+<template>
+  <div class='shuttle'>
+
+    <div class='available'>
+      <label id='available-label'>{{ availableLabel }}</label>
```
Comment (eleanormal) [inline src/components/BShuttle/BShuttle.vue]: This `id` won't be unique on the page if there are 2 shuttles. There's a vue attribute you can use, I think, that's a unique identifier we could append to this. https://stackoverflow.com/questions/34950867/vue-js-how-to-set-a-unique-id-for-each-component-instance

File: src/components/BShuttle/BShuttle.vue (hunk 4 above)
Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: i don't think we want the placeholder to ever be selected, so I'm going to but it on the ul

File: src/components/BShuttle/BShuttle.vue (hunk 5 above)
Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: 👍

File: src/components/BSelect/BSelect.md [hunk 6]
```diff
@@ -1,6 +1,6 @@
 ```js
 
-const options = [{id: 1, displayName: 'This'}, {id: 2, displayName: 'Is'}, {id: 3, displayName: 'An Example'}]
+const options = [{id: 1, displayText: 'This'}, {id: 2, displayText: 'Is'}, {id: 3, displayText: 'An Example'}]
```
Comment (stevenmwade) [inline src/components/BSelect/BSelect.md]: These seem to belong in a different PR

File: src/components/BShuttle/BShuttle.vue [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+          @keydown.space.prevent.stop
+          role='option'
+          tabindex='0'
+        >
+        {{ option.displayText }}
+        </li>
+      </ul>
+    </div>
+
+    <div class='buttons'>
```
Comment (stevenmwade) [inline src/components/BShuttle/BShuttle.vue]: Can we make these less generic or add `b-` to the front of them? This could cause people to accidentally change styles.

File: src/components/BShuttle/BShuttle.vue [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+
+    <div class='chosen'>
+      <label id='chosen-label'>{{ chosenLabel }}</label>
+      <ul
+        aria-labeledby='chosen-label'
+        role='selectbox'
+      >
+        <li
+          v-for="option in chosenOptions"
+          :key="'chosen-option-' + id + '-' + option.id"
```
Comment (stevenmwade) [inline src/components/BShuttle/BShuttle.vue]: Can we also make these raw strings and not do string concatenation?

File: src/components/BShuttle/BShuttle.vue [hunk 9]
```diff
@@ -0,0 +1,480 @@
...
+  
+</template>
+
+<script>
+
+import validationMixIn from '../../mixins/validation'
+
+export default {
+  name: 'b-shuttle',
+  mixins: [validationMixIn()],
```
Comment (stevenmwade) [inline src/components/BShuttle/BShuttle.vue]: You have to call the function? That doesn't seem right according to their docs.

File: src/components/BShuttle/BShuttle.vue [hunk 10]
```diff
@@ -0,0 +1,480 @@
...
+  props: {
+    /**
```
Comment (stevenmwade) [inline src/components/BShuttle/BShuttle.vue]: Are these huge comments what the style guide requires?

File: src/components/BShuttle/BShuttle.vue [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+      selectedOptions: {'available': new Array, 'chosen': new Array},
+      chosenOptions: (this.value && this.options) ? this.setOptions(this.options, this.value, 'chosen') : new Array,
+      lastClick: 0,
+      id: null,
+      allToChosen: require('../../../static/doubleArrow.svg'),
+      allToAvailable: require('../../../static/doubleArrow-left.svg')
+    }
+  },
+
+  computed: {
```
Comment (stevenmwade) [inline src/components/BShuttle/BShuttle.vue]: Don't need this.

File: src/components/BShuttle/BShuttle.vue [hunk 12]
```diff
@@ -0,0 +1,480 @@
...
+  },
+
+  created() {
+    this.$validator.extend('notEmpty', {
+      getMessage: field => 'The ' + field + ' input is required.',
```
Comment (stevenmwade) [inline src/components/BShuttle/BShuttle.vue]: Raw strings

File: src/components/BShuttle/BShuttle.vue [hunk 13]
```diff
@@ -0,0 +1,480 @@
...
+    value: {
+      type: Array,
+      required: false,
+      default: function () { return [] }
+    }
+  },
+
+  data () {
+    return {
+      availableOptions: (this.options && this.options.length > 0) ? this.setOptions(this.options, this.value, 'available') : new Array,
```
Comment (stevenmwade) [inline src/components/BShuttle/BShuttle.vue]: `new Array` instead of `[]`?

File: src/components/BShuttle/BShuttle.vue [hunk 14]
```diff
@@ -0,0 +1,480 @@
...
+     * @param {String} type
+     * this method sets up the initial state of the options
+     */
+    setOptions: function (options, selected, type) {
+      const availableOptions = new Array
+      const chosenOptions = new Array
+      options.forEach(function (option) {
+        if (selected === null) {
+          availableOptions.push(option)
+        } else if (selected.indexOf(option.id) === -1) {
```
Comment (stevenmwade) [inline src/components/BShuttle/BShuttle.vue]: Making this an `||` since they have the same outcome

File: src/components/BShuttle/BShuttle.vue (hunk 7 above)
Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: thats fair, will do

File: src/components/BShuttle/BShuttle.vue (hunk 8 above)
Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: sure

File: src/components/BShuttle/BShuttle.vue (hunk 9 above)
Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: our mixin is a function, not an object

File: src/components/BShuttle/BShuttle.vue (hunk 10 above)
Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: yes

File: src/components/BShuttle/BShuttle.vue (hunk 11 above)
Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: sorry, that was a refactor leftover, will remove

File: src/components/BShuttle/BShuttle.vue [hunk 15]
```diff
@@ -0,0 +1,480 @@
...
+     * this is fired when a line is clicked and updates
+     * selectedOptions accordingly.
+     */
+    onOptionClick (option, optionType, otherOptionType, event) {
+      if (this.selectedOptions[otherOptionType].length > 0) { this.selectedOptions[otherOptionType] = new Array }
+      if (event.shiftKey && (this.lastClick > 0)) {
+        this.multipleSelect(option, optionType)
+      } else if (event.altKey) {
+        this.selectedOptions[otherOptionType] = new Array
+        this.selectedOptions[optionType] = [option.id]
```
Comment (stevenmwade) [inline src/components/BShuttle/BShuttle.vue]: I personally don't like the `new Array` syntax, but if we do use that, then this line should be consistent with that and be `new Array(option.id)`

File: src/components/BShuttle/BShuttle.vue [hunk 16]
```diff
@@ -0,0 +1,480 @@
...
+          @keydown.up.prevent.stop
+          @keydown.down.prevent.stop
+          @keydown.space.prevent.stop
+          role='option'
+          tabindex='0'
+        >
+          {{ option.displayText }} 
+
+        </li>
+      <span v-show="errors.has(validationId)" class="error-text" :id="'error-' + id">{{ errors.first(validationId) }}</span>
```
Comment (stevenmwade) [inline src/components/BShuttle/BShuttle.vue]: Is this span supposed to be in the `ul`? And if so it should be tabbed in

File: src/components/BShuttle/BShuttle.vue (hunk 14 above)
Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: 👍

File: src/components/BShuttle/BShuttle.vue (hunk 15 above)
Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: ill change it

File: src/components/BShuttle/BShuttle.vue (hunk 13 above)
Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: ill change it

File: src/components/BShuttle/BShuttle.vue (hunk 12 above)
Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: 👍

File: src/components/BShuttle/BShuttle.vue (hunk 16 above)
Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: its the error message, it belongs there

File: src/components/BShuttle/BShuttle.vue [hunk 17]
```diff
@@ -0,0 +1,480 @@
...
+     * selectedOptions accordingly.
+     */
+    onOptionClick (option, optionType, otherOptionType, event) {
+      if (this.selectedOptions[otherOptionType].length > 0) { this.selectedOptions[otherOptionType] = new Array }
+      if (event.shiftKey && (this.lastClick > 0)) {
+        this.multipleSelect(option, optionType)
+      } else if (event.altKey) {
+        this.selectedOptions[otherOptionType] = new Array
+        this.selectedOptions[optionType] = [option.id]
+        this.onMoveSelectedClick(optionType)
```
Comment (stevenmwade) [inline src/components/BShuttle/BShuttle.vue]: So this is calling an event function that this is not attached to. This seems like there should be a move function that takes a set of options. The current move function is dependent on an outside variable and it causes this section to immediately select it then move it.

File: src/components/BShuttle/BShuttle.vue [hunk 18]
```diff
@@ -0,0 +1,480 @@
...
+      } else {
+        this.availableOptions.push.apply(this.availableOptions, availableOptions)
+        this.sortById(this.availableOptions)
+        this.chosenOptions = this.sortById(chosenOptions)
+      }
+      this.selectedOptions = {'available': new Array, 'chosen': new Array}
+      if (this.required) {
+        this.validate()
+      }
+      this.$emit("input",Array.from(this.chosenOptions, option => option.id));
```
Comment (stevenmwade) [inline src/components/BShuttle/BShuttle.vue]: Single quotes

File: src/components/BShuttle/BShuttle.vue [hunk 19]
```diff
@@ -0,0 +1,480 @@
...
+      this.selectOption(option.id, optionType)
+    },
+
+    /**
+     * @param {Array<Object>} options
+     * @param {String} type
+     * this moves all selected options from available to 
+     * chosen, or back depending on type.
+     */
+    onMoveSelectedClick (type) {
```
Comment (stevenmwade) [inline src/components/BShuttle/BShuttle.vue]: I think these functions would be easier to read if they were separate functions for available and chosen options. There are a lot of if statements making it a bit challenging to read.

File: src/components/BShuttle/BShuttle.vue (hunk 16 above)
Comment (stevenmwade) [reply src/components/BShuttle/BShuttle.vue]: It belongs in the unordered list? `span` is technically not a valid child element of a `ul`

File: src/components/BShuttle/BShuttle.vue (hunk 17 above)
Comment (stevenmwade) [reply src/components/BShuttle/BShuttle.vue]: This didn't actually change. Not sure why it's outdated

File: src/components/BShuttle/BShuttle.vue (hunk 19 above)
Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: working on it

File: src/components/BShuttle/BShuttle.vue (hunk 16 above)
Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: i like the user experience with it

File: build/webpack.base.config.js [hunk 20]
```diff
@@ -21,5 +21,6 @@ module.exports = {
     new FriendlyErrorsWebpackPlugin({
       clearConsole: true
     })
-  ]
+  ],
```
Comment (eleanormal) [inline build/webpack.base.config.js]: I thought we decided against trailing commas?

File: src/components/BShuttle/BShuttle.vue (hunk 4 above)
Comment (eleanormal) [reply src/components/BShuttle/BShuttle.vue]: The only reason I would have chosen the placeholder over the `ul` is that we don't want the `ul` to have focus when there are options so the tabindex would need to be based on whether the options are shown or not.

File: src/components/BShuttle/BShuttle.vue (hunk 16 above)
Comment (eleanormal) [reply src/components/BShuttle/BShuttle.vue]: So does this show within the shuttle instead of below it? It would be more consistent with other errors if it showed below.

Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: fine

File: src/components/BShuttle/BShuttle.vue [hunk 21]
```diff
@@ -0,0 +1,486 @@
...
+      this.$emit("input",Array.from(this.chosenOptions, option => option.id));
+    },
+
+    /**
+     * @param {String} type
+     * This will move all options to chosen or available
+     * depending on tyep
+     */
+    onMoveAllOptions (type) {
+      if (type === 'available') {
```
Comment (stevenmwade) [inline src/components/BShuttle/BShuttle.vue]: Looks like you could also use `moveItems` here also.

File: src/components/BShuttle/BShuttle.vue [hunk 22]
```diff
@@ -0,0 +1,486 @@
...
+      } else if (event.altKey) {
+        if (optionType === 'available') {
+          const index = this.availableOptions.findIndex(item => item.id === option.id)
+          const optionToMove = this.availableOptions[index]
+          this.availableOptions.splice(index, 1)
+          this.moveItems(this.availableOptions, [optionToMove], optionType)
+        } else {
+          const index = this.chosenOptions.findIndex(item => item.id === option.id)
+          const optionToMove = this.chosenOptions[index]
+          this.chosenOptions.splice(index, 1)
```
Comment (stevenmwade) [inline src/components/BShuttle/BShuttle.vue]: This could be a nice as callback for `moveItems`

File: src/components/BShuttle/BShuttle.vue [hunk 23]
```diff
@@ -0,0 +1,486 @@
...
+      if (optionType === 'available') {
+        this.availableOptions = this.sortById(moveFrom)
+        this.chosenOptions.push.apply(this.chosenOptions, moveTo)
+        this.sortById(this.chosenOptions)
+      } else {
+        this.availableOptions.push.apply(this.availableOptions, moveTo)
+        this.sortById(this.availableOptions)
+        this.chosenOptions = this.sortById(moveFrom)
+      }
+      this.selectedOptions = {'available': [], 'chosen': []}
```
Comment (stevenmwade) [inline src/components/BShuttle/BShuttle.vue]: Could be the same callback also

File: src/components/BShuttle/BShuttle.vue [hunk 24]
```diff
@@ -0,0 +1,490 @@
...
+        >
+        {{ option.displayText }}
+        </li>
+      </ul>
+    </div>
+
+    <div class='b-shuttle-buttons'>
+      <button 
+        @click="onMoveAllOptions('available')"
+        aria-label="Move All To Chosen"
```
Comment (eleanormal) [inline src/components/BShuttle/BShuttle.vue]: All these buttons should have a title attribute instead of aria-label so that the title shows up on hover

Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: 👍

File: src/components/BShuttle/BShuttle.vue (hunk 22 above)
Comment (mlear) [reply src/components/BShuttle/BShuttle.vue]: what benefit would we gain from that?

File: dev/Playground.vue [hunk 25]
```diff
@@ -1,49 +1,14 @@
 <template>
-    <div>
```
Comment (kierachell) [inline dev/Playground.vue]: want me to add a cleanup script to the precommit?

This will wipe the `Playground.vue` file

Comment (mlear) [reply dev/Playground.vue]: i don't think I would want it wiped on every commit, that might get annoying

File: assetsTransformer.js [hunk 26]
```diff
@@ -0,0 +1,7 @@
+const path = require('path');
+
+module.exports = {
+  process(src, filename, config, options) {
+    return 'module.exports = ' + JSON.stringify(path.basename(filename)) + ';';
+  },
```
Comment (eleanormal) [inline assetsTransformer.js]: Is this trailing comma intentional?

Comment (mlear) [reply assetsTransformer.js]: no... i swear I didn't put that there
